docs(tutorial): add new tutorial for serving custom models in Instill Core (#877)
instill-ai/instill.tech: tutorials/model-serving-on-instill-core.mdx

## Why use ⚗️ Instill Model?

1. Seamless Integration with **💧 Instill VDP**: integrate effortlessly with our
   Versatile Data Pipeline, allowing for streamlined unstructured data ETL and
   model serving workflows.
2. No Code Console Builder: Easily utilize custom models defined with **⚗️
   Instill Model** as modular AI components via **📺 Instill Console**, allowing
   for seamless integration into downstream tasks.
3. AutoML Feature (Coming Soon): With the upcoming AutoML feature, **⚗️ Instill
   Model** will soon be capable of automating model training and tuning,
   simplifying model optimization for deployment.

This step-by-step tutorial will guide you through the process of setting up your
own custom model with **Instill Model** for local deployment with **🔮 Instill
Core**.

## Prerequisites

1. **Docker**: Both **🔮 Instill Core** and **⚗️ Instill Model** use **Docker**
   to ensure that models and code can be deployed in consistent, isolated and
   reproducible environments. Please ensure that you have Docker installed and
   running by following the [official
   instructions](https://docs.docker.com/get-docker/), and see our [deployment

>    guide](./docs/core/deployment) for recommended resource settings.

Reviewer: ```suggestion
   guide](/docs/core/deployment) for recommended resource settings.
```

2. **⌨️ Instill CLI**: The easiest way to launch **🔮 Instill Core** for local
   deployment is via the **⌨️ Instill CLI**. To install **⌨️ Instill CLI** using
   Homebrew, please run the following command in your terminal:
   ```bash
   brew install instill-ai/tap/inst
   ```
3. **Launch 🔮 Instill Core**: To launch simply run:
   ```bash
   inst local deploy
   ```
   Please note that the initial launch process may take up to 1 hour, depending
   on your internet speed. Subsequent launches will be much faster, usually
   completing in under 5 minutes.
4. Now that **🔮 Instill Core** has been deployed, you can access **📺 Instill
   Console** at [http://localhost:3000](http://localhost:3000). Please use the
   following initial login details to initiate the password reset process for
   onboarding:

    - Username: `admin`
    - Password: `password`

For further details about launching **🔮 Instill Core**, we recommend that you

> refer to the [deployment guide](./docs/core/deployment) which covers alternative

Reviewer: ```suggestion
refer to the [deployment guide](/docs/core/deployment) which covers alternative
```

ways you can launch with Docker Compose or Kubernetes with Helm.

Finally, please note that this guide assumes you have a basic understanding of
machine learning and can code in Python. If you are new to these concepts, we
recommend that you look at our [quick start
guide](/docs/latest/quickstart#authorisation?utm_source=tutorial&utm_medium=link&utm_campaign=model-oss-llm)
which introduces our no-code **💧 Instill VDP** pipeline builder, and also take
a look at some of our other [tutorials](https://www.instill.tech/tutorials).

# Step-by-Step Tutorial

## Step 1: Create a Model Namespace

To get started, navigate to the Model page in the Console window and click the
**+ Create Model** button.

This should bring up a configuration window (see image below) where you are able
to configure your model settings. For a full description of the available
fields, please refer to the [Create a Model](/docs/model/create) page.

In this tutorial, we will be walking through how to create and deploy a version
of the
[TinyLlama-1.1B-Chat](https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v1.0)
model. To follow along, please fill in the configuration fields as per the image
below.

> <ZoomableImg src="/tutorial-assets/model-serving-on-instill-core/T1.png" alt="2.
> Configure Model Settings" />

Reviewer: ```suggestion
<ZoomableImg src="/tutorial-assets/model-serving-on-instill-core/T1.png" alt="Configure Model Settings" />
```

You have now created an empty model namespace on **⚗️ Instill Model**. In the
next sections of this tutorial we will show you how to define your own custom
model for deployment!

## Step 2: Create a Model Config

To prepare a model to be served with **⚗️ Instill Model**, you first need to
create your own model directory containing two files - `model.py` and
`instill.yaml`.

To configure the
[TinyLlama-1.1B-Chat](https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v1.0)
model, simply open the `instill.yaml` file and populate it with:

<CH.Code>
```yaml instill.yaml
build:
  gpu: true
  python_version: "3.11"
  python_packages:
     - torch==2.2.1
     - transformers==4.36.2
     - accelerate==0.25.0

> repo: {user_id}/{model_id}

Reviewer: ```suggestion
repo: USER_ID/MODEL_ID
```

```
</CH.Code>

> Importantly, you must replace `{user_id}` with `admin`, and replace `{model_id}`

Reviewer: ```suggestion
Importantly, you must replace `USER_ID` with `admin`, and replace `MODEL_ID`
```

, to align with the [doc](https://www.instill.tech/docs/model/prepare#model-configuration).

from transformers import pipeline

from instill.helpers.ray_config import instill_deployment, InstillDeployable

@instill_deployment
class TinyLlama:
    def __init__(self):
        self.pipeline = pipeline(
            "text-generation",
            model="TinyLlama/TinyLlama-1.1B-Chat-v1.0",
            torch_dtype=torch.bfloat16,
            device_map="auto",
        )
```
</CH.Code>

The `instill.helpers.ray_config` package contains the decorators and deployment
object for the model class, which we will use to convert the model class into a
servable model. These are required to properly define a model class for **⚗️
Instill Model**.

### Define the Model Metadata method

In the second phase, we define the `ModelMetadata` method which is responsible
for communicating the models expected input and output shapes to the backend
service. To easily facilitate this, we can make use of the [Python
SDK](/docs/latest/sdk/python-sdk) through the `instill.helpers` module which
provides a number of functions that can be selected according to the AI task the
model performs.

Here, we recognise that the
[TinyLlama-1.1B-Chat](https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v1.0)

> model falls under the `text_generation_chat` AI task, and so we will make use of

Reviewer: ```suggestion
model falls under the [Text Generation Chat](https://www.instill.tech/docs/model/ai-task#text-generation-chat) AI task, and so we will make use of
```
, to cite the AI task reference.

the `construct_text_generation_chat_metadata_response` helper function.

Please refer [here](https://www.instill.tech/docs/model/ai-task) for a full list
of the supported AI tasks.

<CH.Code>
```python model.py
# previous imports defined above

from instill.helpers import construct_text_generation_chat_metadata_response

@instill_deployment
class TinyLlama:
    # __init__ defined in phase 1

    def ModelMetadata(self, req):
        return construct_text_generation_chat_metadata_response(req=req)
```
</CH.Code>

### Implement the Inference Method

In the third phase, we implement the inference method `__call__`, which handles
the trigger request from **⚗️ Instill Model**, contains the necessary logic to
run the inference, and constructs the response. We use the `StandardTaskIO`
module to parse the request payload into input parameters, and convert the model
outputs to the appropriate response format.

The `TextGenerationChatInput` class from `instill.helpers.const` is used to

> define the input format for the text generation chat AI task, and the

Reviewer: ```suggestion
define the input format for the [Text Generation Chat](https://www.instill.tech/docs/model/ai-task#text-generation-chat) AI task, and the
```

    async def __call__(self, request):
        # parse the request and get the corresponding input for text-generation-chat task
        task_text_generation_chat_input: TextGenerationChatInput = (
            StandardTaskIO.parse_task_text_generation_chat_input(request=request)
        )

        # prepare prompt with chat template
        conv = [
            {
                "role": "system",
                "content": "You are a friendly chatbot",
            },
            {
                "role": "user",
                "content": task_text_generation_chat_input.prompt,
            },
        ]

        prompt = self.pipeline.tokenizer.apply_chat_template(
            conv,
            tokenize=False,
            add_generation_prompt=True,
        )

        # inference
        sequences = self.pipeline(
            prompt,
            max_new_tokens=task_text_generation_chat_input.max_new_tokens,
            do_sample=True,
            temperature=task_text_generation_chat_input.temperature,
            top_k=task_text_generation_chat_input.top_k,
            top_p=0.95,
        )

        # convert the model output into response output

>         task_text_generation_chat_output = (
>             StandardTaskIO.parse_task_text_generation_chat_output(sequences=sequences)
>         )

Reviewer: ```suggestion
        task_text_generation_chat_output: TextGenerationChatInput = (
            StandardTaskIO.parse_task_text_generation_chat_output(sequences=sequences)
        )
```

I supposed the type hint should work as in the input case?

Author: I don't think so, I think for the output case this returns `serialized_bytes_tensor`.

    async def __call__(self, request):
        # parse the request and get the corresponding input for text-generation-chat task
        task_text_generation_chat_input: TextGenerationChatInput = (
            StandardTaskIO.parse_task_text_generation_chat_input(request=request)
        )

        # prepare prompt with chat template
        conv = [
            {
                "role": "system",
                "content": "You are a friendly chatbot",
            },
            {
                "role": "user",
                "content": task_text_generation_chat_input.prompt,
            },
        ]

        prompt = self.pipeline.tokenizer.apply_chat_template(
            conv,
            tokenize=False,
            add_generation_prompt=True,
        )

        # inference
        sequences = self.pipeline(
            prompt,
            max_new_tokens=task_text_generation_chat_input.max_new_tokens,
            do_sample=True,
            temperature=task_text_generation_chat_input.temperature,
            top_k=task_text_generation_chat_input.top_k,
            top_p=0.95,
        )

        # convert the model output into response output

>         task_text_generation_chat_output = (
>             StandardTaskIO.parse_task_text_generation_chat_output(sequences=sequences)
>         )

Reviewer: ```suggestion
        task_text_generation_chat_output: TextGenerationChatInput = (
            StandardTaskIO.parse_task_text_generation_chat_output(sequences=sequences)
        )
```

Author: I don't think so, I think for the output case this returns `serialized_bytes_tensor`.

Reviewer: Hey @heiruwu, can you help confirm this? There is a type `TextGenerationChatOutput` in the [protogen-python](https://github.com/instill-ai/protogen-python/blob/main/model/model/v1alpha/task_text_generation_chat_pb2.pyi#L89), and I feel the input and output expression here could be symmetric.

Reviewer 2: no, the `task_text_generation_chat_output` here is the serialized bye string that will be passed back to model-backend, not the user.
the `TextGenerationChatOutput` class is the response for the trigger endpoint.

Reviewer 2: the `TextGenerationChatInput` class is written [here](https://github.com/instill-ai/python-sdk/blob/main/instill/helpers/const.py#L58-L68) for the user's convenience to understand the request data structure after its been deserialized, not by the protogen-go

Author: @pinglin @heiruwu are we ok to resolve and merge?

with **⚗️ Instill Model**. In the next step, we will show you how to build and
deploy this model locally with **🔮 Instill Core**!

## Step 4: Build and Deploy the Model

First, you must ensure that you have the same Python version installed in your
local environment as specified in the `instill.yaml` file in [step
2](#step-2-create-a-model-config), in this case `python_version: "3.11"`.

You then need to ensure you have installed the latest version of the [Python
SDK](/docs/latest/sdk/python-sdk) by running:

```bash
pip install instill-sdk
```

### Build the Model Image

You can now build your model image by running the following command from within
the directory containing the `model.py` and `instill.yaml` files:

```bash
instill-build -t v1
```

This command will build the model image under version tag `v1`. Upon successful
completion, you should see a similar output to the following:
```
2024-05-28 01:54:44,404.404 INFO     [Instill Builder] admin/tinyllama:v1 built
2024-05-28 01:54:44,423.423 INFO     [Instill Builder] Done
```

### Push the Model Image

To push the model image to the **🔮 Instill Core** instance we will need to be

> able to login to Docker. To do this, you first need to create an API token by:

Reviewer: ```suggestion
able to login to the hosted Docker registry. To do this, you first need to create an API token by:
```

1. Selecting the profile icon in the top right corner of the Console window and
   choosing the Settings option.
2. Select API Tokens from the left-hand menu.
3. Click the **Create Token** button and give it a name, e.g. `tutorial`. Copy
   the generated API token.

> <ZoomableImg src="/tutorial-assets/model-serving-on-instill-core/T2.png" alt="2.
> Create API Token" />

Reviewer: ```suggestion
<ZoomableImg src="/tutorial-assets/model-serving-on-instill-core/T2.png" alt="Create API Token" />
```

> Now we can login to Docker on the **🔮 Instill Core** instance by running:

Reviewer: ```suggestion
Now we can login to the Docker register in the **🔮 Instill Core** instance by running:
```

<ZoomableImg src="/tutorial-assets/model-serving-on-instill-core/T3.png" alt="3.
Check Model Status" />

The `Status` will initially show as `Starting`, indicating that your model is
offline and **⚗️ Instill Model** is still in the process of allocating resources
and deploying it (this may take a few minutes). Once this status changes to
`Active`, your model is ready to serve requests. 🚀

## Step 5. Inference

Once your model is deployed and `Active`, you can easily test its behaviour
following these steps:
1. Navigating to the **Overview** tab for your `admin/tinyllama` model.
2. Enter a prompt in the **Input** pane (e.g. What is a rainbow?)
3. Scroll down and hit **Run** to trigger the model inference.

<ZoomableImg src="/tutorial-assets/model-serving-on-instill-core/T4.png" alt="4.
Model Inference" />

You should see a response generated by the
[TinyLlama-1.1B-Chat](https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v1.0)
model in the **Output** pane.

To access model inferences via the API:
1. Navigate to the **API** tab for your `admin/tinyllama` model.
2. Follow the instructions by setting your own `INSTILL_API_TOKEN` as an
   environment variable:
   ```bash

>    $ export INSTILL_API_TOKEN=********

Reviewer: ```suggestion
   export INSTILL_API_TOKEN=********
```

, to align with the other command line code snippets.
